### PR TITLE
Express Dev Serve

### PR DESCRIPTION
**Pull request by GHaberis**
## What does this change?

Replaces the webpack dev server with an `express` Dev Server used in conjunction with `webpack-hot-server-middleware` and `webpack-dev-middleware`.

## Why?

- hot module replacement for the client using `webpack-hot-middleware`
-  ensure the server bundle used is the latest without requiring a restart using `webpack-hot-server-middleware`

### CODE REVIEW

File: __config__/webpack/webpack.config.dev.js [hunk 1]
```diff
@@ -1,114 +1,50 @@
...
+            `DEV server running at ${chalk.blue.underline(
+                'http://localhost:3000',
+            )}`,
+        ],
+    },
 });
 
 module.exports = {
     browser: {
+        name: 'browser',
```
Comment (sndrs) [inline __config__/webpack/webpack.config.dev.js]: these could be pushed up to the main config? they're handy labels in all envs

File: __config__/webpack/webpack.config.dev.js [hunk 2]
```diff
@@ -1,114 +1,50 @@
...
         devtool: 'cheap-module-eval-source-map',
         entry: {
+            app: ['webpack-hot-middleware/client', './src/browser'],
             demo: './__tools__/demo/demo.browser.js',
             src: './__tools__/demo/src.browser.js',
         },
-        devServer: {
+        output: {
+            path: path.resolve(__dirname, '../../', 'dist'),
```
Comment (sndrs) [inline __config__/webpack/webpack.config.dev.js]: i think this is set in main config already?

File: __config__/webpack/webpack.config.dev.js [hunk 3]
```diff
@@ -1,114 +1,50 @@
...
-                },
-            }),
+            new webpack.HotModuleReplacementPlugin(),
+            new webpack.NamedModulesPlugin(),
+            friendlyErrorsWebpackPlugin,
         ],
     },
-    server: {},
+    server: {
+        name: 'server',
```
Comment (sndrs) [inline __config__/webpack/webpack.config.dev.js]: again, can be pushed up to main config

File: __config__/webpack/webpack.config.dev.js [hunk 4]
```diff
@@ -1,114 +1,50 @@
...
+        target: 'node',
```
Comment (sndrs) [inline __config__/webpack/webpack.config.dev.js]: i think this is set already?

Comment (sndrs) [reply __config__/webpack/webpack.config.dev.js]: i'm going to add a few "i think this is set already?" comments, sorry for the spam :)

File: __config__/webpack/webpack.config.dev.js [hunk 5]
```diff
@@ -1,114 +1,50 @@
...
+        entry: {
+            app: './src/server',
```
Comment (sndrs) [inline __config__/webpack/webpack.config.dev.js]: i think this is set already?

File: __config__/webpack/webpack.config.dev.js [hunk 6]
```diff
@@ -1,114 +1,50 @@
...
+            demo: './__tools__/demo/demo.browser.js',
+            src: './__tools__/demo/src.browser.js',
+        },
+        output: {
+            path: path.resolve(__dirname, '../../', 'dist'),
```
Comment (sndrs) [inline __config__/webpack/webpack.config.dev.js]: i think this is set already?

File: __config__/webpack/webpack.config.dev.js [hunk 7]
```diff
@@ -1,114 +1,50 @@
...
+            libraryTarget: 'commonjs2',
```
Comment (sndrs) [inline __config__/webpack/webpack.config.dev.js]: i think this need to be moved up for dev and prod envs?

File: __tools__/dev-server.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        compiler.compilers.find(config => config.name === 'browser'),
+        {
+            // https://www.npmjs.com/package/friendly-errors-webpack-plugin#turn-off-errors
+            log: () => {},
+        },
+    ),
+);
+
+require('@babel/register')({
+    only: [/__tools__\/demo|src/],
```
Comment (sndrs) [inline __tools__/dev-server.js]: only needs to apply to `__tools__` now?

File: src/server.js [hunk 9]
```diff
@@ -14,3 +16,17 @@ export default async (state: { page: string }): string => {
 
     return doc({ html, state, css, cssIDs });
 };
+
+// eslint-disable-next-line no-unused-vars
```
Comment (sndrs) [inline src/server.js]: can just not have the `next` arg?

File: src/server.js [hunk 10]
```diff
@@ -14,3 +16,17 @@ export default async (state: { page: string }): string => {
...
+export default () => (req, res, next) => {
+    const pageType = req.params[0].split('/pages/')[0];
```
Comment (sndrs) [inline src/server.js]: should this be in `src`?

File: __config__/webpack/webpack.config.dev.js (hunk 4 above)
Comment (GHaberis) [reply __config__/webpack/webpack.config.dev.js]: No worries, I'll do a cross check with the main config now and remove anything that's specified there from this config!

File: src/server.js [hunk 11]
```diff
@@ -1,11 +1,13 @@
...
+/* eslint-disable global-require,import/no-dynamic-require */
 
+import { log } from 'util';
 import { renderToString } from 'react-dom/server';
 import { extractCritical } from 'emotion-server';
 
 import doc from 'lib/__html';
 
-export default async (state: { page: string }): string => {
+const fetchPage = async (state: { page: string }) => {
```
Comment (SiAdcock) [inline src/server.js]: Lost our return type here!

File: src/server.js (hunk 9 above)
Comment (GHaberis) [reply src/server.js]: Yes `next` can be removed as an argument

File: __config__/webpack/webpack.config.dev.js [hunk 12]
```diff
@@ -1,114 +1,50 @@
...
         ],
     },
-    server: {},
+    server: {
+        name: 'server',
+        target: 'node',
+        entry: {
+            app: './src/server',
+            demo: './__tools__/demo/demo.browser.js',
+            src: './__tools__/demo/src.browser.js',
```
Comment (SiAdcock) [inline __config__/webpack/webpack.config.dev.js]: Should these point to server entry points (`demo.server.js` and `src.server.js`)?

Comment (GHaberis) [reply __config__/webpack/webpack.config.dev.js]: hah, yes well spotted!

File: src/server.js [hunk 13]
```diff
@@ -14,3 +16,17 @@ export default async (state: { page: string }): string => {
...
+    const data = require(`../.data/${pageType}`);
+
+    fetchPage(data)
+        .then(html => {
+            res.status(200).send(html);
+        })
+        .catch(e => {
+            log(e);
+        });
+};
```
Comment (SiAdcock) [inline src/server.js]: Could we `async await` this up?

```js
try {
  const html = await fetchPage(data)
  res.status(200).send(html)
} catch (e) {
  log(e)
})
```

File: src/server.js [hunk 14]
```diff
@@ -1,11 +1,13 @@
 // @flow
+/* eslint-disable global-require,import/no-dynamic-require */
 
+import { log } from 'util';
```
Comment (SiAdcock) [inline src/server.js]: NB: this method is [deprecated](https://nodejs.org/docs/latest-v8.x/api/util.html#util_util_log_string). I have used [`bunyan`](https://github.com/trentm/node-bunyan) which did what we needed, but perhaps there are better alternatives these days.

File: __tools__/dev-server.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+});
+
+app.get('/src/*', async (req, res) => {
+    const src = require('./demo/src.server').default;
+    try {
+        res.send(src(req.params[0]));
+    } catch (e) {
+        log(e);
+    }
+});
```
Comment (SiAdcock) [inline __tools__/dev-server.js]: Do our route handlers need to be `async`? It doesn't look like there's any `await`ing happening

File: __config__/webpack/webpack.config.dev.js [hunk 16]
```diff
@@ -1,114 +1,40 @@
...
     browser: {
         devtool: 'cheap-module-eval-source-map',
         entry: {
+            app: ['webpack-hot-middleware/client', './src/browser'],
             demo: './__tools__/demo/demo.browser.js',
             src: './__tools__/demo/src.browser.js',
         },
-        devServer: {
+        output: {
             publicPath: '/assets/javascript/',
```
Comment (sndrs) [inline __config__/webpack/webpack.config.dev.js]: should this be in top level?

**Pull request merged**